### PR TITLE
Add map and navigation bar to home

### PR DESCRIPTION
**Pull request by raul0208**
Add Navigation Bar and Map with annotation to home screen.

Video:

https://user-images.githubusercontent.com/112905642/197006478-c8378338-bc05-427b-8459-068bac57d715.mov

### CODE REVIEW

File: ios-target-app/Home/Views/HomeViewController.swift [hunk 1]
```diff
@@ -6,13 +6,107 @@
...
+      longitudinalMeters: UI.MapRegion.longitudinalMeters
+    )
+    mapView.setRegion(viewRegion, animated: false)
+  }
+  
+  private func setMapConstraints() {
+    view.addSubview(mapView)
+    
+    mapView.translatesAutoresizingMaskIntoConstraints = false
+
```
Comment (tarruk) [inline ios-target-app/Home/Views/HomeViewController.swift:106]: when you have more than one Constraints you can create an array of constraints and activate all of them at the same time, something like this: 

```swift
NSLayoutConstraint.activate([
  mapView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
  mapView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor),
  mapView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
  mapView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor),
])
```

File: ios-target-app/Managers/LocationManager.swift [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+//  LocationManager.swift
+//  ios-target-app
+//
+//  Created by Raul Piñeres Carrera on 13/10/22.
+//
+
+import UIKit
+import CoreLocation
+
+class LocationManager: NSObject, CLLocationManagerDelegate {
```
Comment (tarruk) [inline ios-target-app/Managers/LocationManager.swift:11]:
```suggestion
final class LocationManager: NSObject, CLLocationManagerDelegate {
```

File: ios-target-app/Home/Views/HomeViewController.swift [hunk 3]
```diff
@@ -6,13 +6,107 @@
...
+    map.overrideUserInterfaceStyle = .dark
+    map.isZoomEnabled = true
+    map.isScrollEnabled = true
+    return map
+  }()
+  
+  private let locationManager: LocationManager
+  private var subscribers = Set<AnyCancellable>()
+  
+  init(locationManager: LocationManager) {
```
Comment (glm4) [inline ios-target-app/Home/Views/HomeViewController.swift:25]: Have you considered using a different design approach here? Maybe MVVM or similar

File: ios-target-app/Home/Views/HomeViewController.swift [hunk 4]
```diff
@@ -6,13 +6,107 @@
...
+  }
+  
+  override func viewWillAppear(_ animated: Bool) {
+    super.viewWillAppear(animated)
+    mapView.showsUserLocation = true
+  }
+  
+  override func viewWillDisappear(_ animated: Bool) {
+    super.viewWillDisappear(animated)
+    mapView.showsUserLocation = false
```
Comment (glm4) [inline ios-target-app/Home/Views/HomeViewController.swift:58]: what is the intention for this to be set on the viewWillAppear and viewWillDisappear?

File: ios-target-app/Home/Views/HomeViewController.swift [hunk 5]
```diff
@@ -6,13 +6,107 @@
...
+  
+  private func setMapConstraints() {
+    view.addSubview(mapView)
+    
+    mapView.translatesAutoresizingMaskIntoConstraints = false
+    
+    mapView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor).isActive = true
+    mapView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor).isActive = true
+    mapView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor).isActive = true
+    mapView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor).isActive = true
```
Comment (glm4) [inline ios-target-app/Home/Views/HomeViewController.swift:110]: You could use `NSLayoutConstraint.activate([...])`
to activate multiple constraints at the same time

File: ios-target-app/Managers/LocationManager.swift [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  @Published var lastLocation: CLLocation?
+  
+  override init() {
+    super.init()
+    locationManager.delegate = self
+    locationManager.desiredAccuracy = kCLLocationAccuracyBest
+    
+    if CLLocationManager.locationServicesEnabled() { // Check for Location Services
+      locationManager.requestWhenInUseAuthorization()
+      locationManager.startUpdatingLocation()
```
Comment (glm4) [inline ios-target-app/Managers/LocationManager.swift:25]: this IMO should be done in a public function that you expose, and consumers of LocationManager decide when to start/stop listening for location updates.
On a different topic, this should also check for the user permissions, I think `locationServicesEnabled` doesn't take that into account.

File: ios-target-app/Home/Views/HomeViewController.swift [hunk 7]
```diff
@@ -6,13 +6,107 @@
...
+      action: nil
+    )
+    
+    profileButtonIcon.tintColor = .black
+    chatButtonIcon.tintColor = .black
+    
+    navigationItem.leftBarButtonItem = profileButtonIcon
+    navigationItem.rightBarButtonItem = chatButtonIcon
+  }
+  private func setupSubscribers() {
```
Comment (kstoletniy) [inline ios-target-app/Home/Views/HomeViewController.swift:84]:
```suggestion
  }
  
  private func setupSubscribers() {
```

File: ios-target-app/Managers/LocationManager.swift (hunk 6 above)
Comment (raul0208) [reply ios-target-app/Managers/LocationManager.swift:25]: mmm Ok this one only checks if it is enabled or not, right? but it is not that accurate.

File: ios-target-app/Home/Views/HomeViewController.swift (hunk 3 above)
Comment (raul0208) [reply ios-target-app/Home/Views/HomeViewController.swift:25]: You mean like addressing this inside a ViewModel instead?

File: ios-target-app/Home/Views/HomeViewController.swift (hunk 4 above)
Comment (glm4) [reply ios-target-app/Home/Views/HomeViewController.swift:58]: For now is ok then, later you will use a custom map annotation for the user location.

File: ios-target-app/Home/ViewModels/HomeViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  private let locationManager: LocationManager
+  
+  init(locationManager: LocationManager = LocationManager.instance) {
+    self.locationManager = locationManager
+    
+    setupSubscribers()
+  }
+  
+  @Published private var currentLocation: CLLocation = CLLocation()
+  var currentLocationPublisher: Published<CLLocation>.Publisher { $currentLocation }
```
Comment (germantw) [inline ios-target-app/Home/ViewModels/HomeViewModel.swift:25]: do you really need the currentLocation variable?
You can directly pass the signal from the location manager:
```suggestion
  var currentLocationPublisher: Published<CLLocation>.Publisher { 
    locationManager.$lastLocation
  }
```

File: ios-target-app/Home/Views/HomeViewController.swift [hunk 9]
```diff
@@ -6,13 +6,115 @@
...
+  
+  private lazy var mapView: MKMapView = {
+    let map = MKMapView()
+    map.overrideUserInterfaceStyle = .dark
+    map.isZoomEnabled = true
+    map.isScrollEnabled = true
+    return map
+  }()
+  
+  // private let locationManager: LocationManager
```
Comment (germantw) [inline ios-target-app/Home/Views/HomeViewController.swift:24]: pls remove if not necessary

File: ios-target-app/Home/Views/HomeViewController.swift [hunk 10]
```diff
@@ -6,13 +6,115 @@
...
+  private var subscribers = Set<AnyCancellable>()
```
Comment (germantw) [inline ios-target-app/Home/Views/HomeViewController.swift:25]:
```suggestion
  private var cancellableSubscriptions = Set<AnyCancellable>()
```

File: ios-target-app/Managers/LocationManager.swift [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+      case .restricted, .denied:
+        locationManager.stopUpdatingLocation()
+        break
+      case .notDetermined:
+        manager.requestWhenInUseAuthorization()
+        break
+      default:
+        break
+      }
+    }
```
Comment (germantw) [inline ios-target-app/Managers/LocationManager.swift:41]:
```suggestion
    guard CLLocationManager.locationServicesEnabled() else {
      return
    }
    switch manager.authorizationStatus {
    case .authorizedAlways, .authorizedWhenInUse:
      locationManager.startUpdatingLocation()
    case .restricted, .denied:
      locationManager.stopUpdatingLocation()
    case .notDetermined:
      manager.requestWhenInUseAuthorization()
    default:
      break
    }
```

Comment (raul0208) [reply ios-target-app/Managers/LocationManager.swift:41]: Oh, didn't know that. Thanks!

File: ios-target-app/Home/ViewModels/HomeViewModel.swift (hunk 8 above)
Comment (raul0208) [reply ios-target-app/Home/ViewModels/HomeViewModel.swift:25]: gotcha

**Pull request merged**